pkg/attachutils/simpleattach.go

```diff
@@ -0,0 +1,385 @@
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/docker/docker/api/server/httputils"
+	"github.com/docker/docker/pkg/ioutils"
+)
+
+// HijackClosure function type returned from HijackConnection().  The call can
+// later be used to return stdio streams for the connection.
+type HijackClosure func() (io.ReadCloser, io.Writer, io.Writer, error)
+
+// AttachToDockerLikeServer is a utility function that assumes the server that
+// is being attempted for attach follows a URL path along the line of
+// /containers/{containerID}/attach?
+func AttachToDockerLikeServer(baseURL, containerID, attachVerb string, useStdin, useStdout, useStderr, stream bool) (net.Conn, *bufio.Reader, error) {
+	var options map[string]string
+	var buf bytes.Buffer
+
+	// Build up the request endpoint with params from ContainerAttachConfig
+	options = make(map[string]string)
+
+	if useStdin {
+		options["stdin"] = "1"
+	}
+
+	if useStdout {
+		options["stdout"] = "1"
+	}
+
+	if useStderr {
+		options["stderr"] = "1"
+	}
+
+	if stream {
+		options["stream"] = "1"
+	}
+
+	for key, value := range options {
+		if buf.Len() > 0 {
+			buf.WriteString("&")
+		}
+		buf.WriteString(key)
+		buf.WriteString("=")
+		buf.WriteString(value)
+	}
+
+	endpoint := baseURL + "/containers/" + containerID + "/attach?" + buf.String()
+
+	return AttachToServer(endpoint)
+}
+
+// AttachToServer connects to a server, given the address in 'server' and returns
+// a net conn and buffered reader.
+func AttachToServer(server string) (net.Conn, *bufio.Reader, error) {
+	method := "POST"
+
+	// Create a connection client and request
+	c, err := net.DialTimeout("tcp", server, time.Duration(10*time.Second))
+	if err != nil {
+		retErr := fmt.Errorf("could not dial VIC portlayer for attach: %v", err)
+		log.Errorf(retErr.Error())
+		return nil, nil, retErr
+	}
+
+	if tcpConn, ok := c.(*net.TCPConn); ok {
+		tcpConn.SetKeepAlive(true)
+		tcpConn.SetKeepAlivePeriod(30 * time.Second)
+	}
+
+	client := httputil.NewClientConn(c, nil)
+	defer client.Close()
+
+	req, err := http.NewRequest(method, "http://"+server, nil)
+	if err != nil {
+		retErr := fmt.Errorf("could not create request for VIC portlayer attach: %v", err)
+		log.Errorf(retErr.Error())
+		return nil, nil, retErr
+	}
+
+	//	req.Header.Set("Connection", "Upgrade")
+	//	req.Header.Set("Upgrade", "tcp")
+	req.Header.Set("Content-Type", "text/plain")
+
+	// Make the connection and hijack it
+	client.Do(req)
+	conn, br := client.Hijack()
+
+	return conn, br, nil
+}
+
+// CloseWriter is an interface that implement structs
+// that close input streams to prevent from writing.
+type CloseWriter interface {
+	CloseWrite() error
+}
+
+// Code copied from Docker and modified.  stdin, stdout, stderr are the client
+// (hijacked connection) and conn, bufReader are the server
+func AttachStreamsToConn(openStdin, stdinOnce, tty bool, stdin io.ReadCloser, stdout io.Writer, stderr io.Writer, conn net.Conn, bufReader *bufio.Reader, keys []byte) error {
```

Comment:
This function and the one below it are sufficiently similar that they can be made more generic and condensed into 1 function.